File: old-projects/vector-defence-silverlight-port/src/Sprite.ts
import { BoardElement } from "./BoardElement";
import { Game } from "./Game";

export abstract class Sprite extends BoardElement {
  markedForRemoval = false;
  dx = 0;
  dy = 0;
  radius = 0;
  protected alpha = 1;

  protected constructor() {
    super();
    Game.sprites.push(this);
  }

  protected get isOutsideBounds(): boolean {
    return this.x < 0 || this.y < 0 || this.x > Game.canvasWidth || this.y > Game.canvasHeight;
  }

  abstract render(context: CanvasRenderingContext2D): void;

  abstract animate(): void;

  removeFromCanvas(): void {}

  protected updatePosition(): void {
    this.y += this.dy * Game.currentFrameMultiplier;
    this.x += this.dx * Game.currentFrameMultiplier;
  }
}

File: silverlight-ports/vector-defence/original-csharp-source/Sprite.cs
using System.Windows.Shapes;

namespace VectorDefenceSL
{
    public abstract class Sprite : BoardElement
    {
        public bool MarkedForRemoval;
        public double DX;
        public double DY;
        
        protected Shape mainElement;
        protected double alpha = 1;

        protected Sprite()
        {
            Game.Sprites.Add(this);
        }

        public double Radius { get; protected set; }

        protected bool IsOutsideBounds
        {
            get
            {
                return X < 0 || Y < 0 || X > Game.CanvasWidth || Y > Game.CanvasHeight;
            }
        }

        public abstract void Render();

        public abstract void Animate();

        public virtual void RemoveFromCanvas()
        {
            Game.Children.Remove(mainElement);
        }

        protected void UpdatePosition()
        {
            var ndy = DY * Game.CurrentFrameMultiplier;
            var ndx = DX * Game.CurrentFrameMultiplier;
            Y += ndy; // move
            X += ndx;
        }

        protected void Move(double x, double y)
        {
            mainElement.SetCanvasPosition(x, y);
        }
    }
}
